qiskit/transpiler/passes/optimization/hoare_opt.py

```diff
@@ -214,7 +214,7 @@ def _traverse_dag(self, dag):
             if remove_ctrl:
                 dag.substitute_node_with_dag(node, new_dag)
                 gate = gate.base_gate
-                node.op = gate
+                node.op = gate.to_mutable()
```

Comment:
I think generally this is the safer approach here, but I'm wondering where it's coming from. Are we doing follow up mutation on `node.op` somewhere, I'm just trying to grasph why CI didn't flag this in the earlier singleton PRs.

Reply:
It's very hard to parse, but the next line after this highlighted one is `node.name = ...`, and `DAGOpNode.name` is a property that forwards sets to `DAGOpNode.op.name`, which is where the mutation happens.

Reply:
Ah, I forgot about the forwarding we do from `DAGOpNode`, this makes sense then. Although how it passed unit tests I'm not sure, because this clearly wasn't singleton safe.

Reply:
It's because in practice, all previous `ControlledGate` instances that would end up here would have been created by some `__init__` that did something along the lines of
```python
def __init__(self, ctrl_state=None, *, _base_label=None):
    base = XGate(label=_base_label)
    super().__init__("cx", 2, 0, [], num_ctrl_qubits=1, ctrl_state=ctrl_state, base_gate=base)
```
Before this PR, the `base` would be a mutable instance, because the constructor wouldn't be able to tell that the explicit setting of `label` was just the default.